sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -362,42 +346,46 @@ public void onError(Throwable e) {
                     public void onCompleted() {
                       ReadRowsRequest nextNextRequest = null;
 
-                      if (byteLimitReached || rows.size() == segmentSize) {
+                      // When requested rows < limit, the current request will be the last
+                      if (byteLimitReached || rows.size() == nextRequest.getRowsLimit()) {
                         nextNextRequest =
                             truncateRequest(nextRequest, rows.get(rows.size() - 1).getKey());
                       }
-
                       f.set(new UpstreamResults(rows, nextNextRequest));
                     }
                   });
-      atomic.set(handler);
+      atomicScanHandler.set(handler);
       return f;
     }
 
-    void waitReadRowsFuture() throws IOException {
+    private void waitReadRowsFuture() throws IOException {
       try {
         UpstreamResults r = future.get();
         buffer.addAll(r.rows);
         nextRequest = r.nextRequest;
         future = null;
         serviceCallMetric.call("ok");
-      } catch (StatusRuntimeException e) {
-        serviceCallMetric.call(e.getStatus().getCode().value());
-        throw e;
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
         throw new IOException(e);
       } catch (ExecutionException e) {
-        throw new IOException(e);
+        throw new IOException(e.getCause());
+      }
+      catch (Exception e) {
+        if (e instanceof InterruptedException) {
```

Comment:
Should this also be `if (e.getCause() instanceof InterruptedException)`?

Reply:
I meant to delete the catch (Exception e) for the catches above. Is there a need for checking e.getCause()?